README.md
[![Build Status](https://travis-ci.com/asalamon74/pktriggercord.svg?branch=master)](https://travis-ci.com/asalamon74/pktriggercord)
[![GitHub release](https://img.shields.io/github/release/asalamon74/pktriggercord.svg)](https://github.com/asalamon74/pktriggercord/releases)
[![Packaging status](https://repology.org/badge/tiny-repos/pktriggercord.svg)](https://repology.org/metapackage/pktriggercord/versions)
[![CodeFactor](https://www.codefactor.io/repository/github/asalamon74/pktriggercord/badge)](https://www.codefactor.io/repository/github/asalamon74/pktriggercord)

# PkTriggerCord

Remote control (tethering) for Pentax DSLR cameras.

After you connect the camera to your PC with the USB cable, you can
control the camera from pkTriggerCord. You can read status
information, you can change several settings, and you can take
pictures. The pictures are stored in the camera's buffer and can be
downloaded to the PC. Pictures are not written on the SD card of
the camera.

## Supported systems

- Linux x86
- Linux arm (Raspberry PI)
- OpenBSD
- Windows
- [Android](http://pktriggercord.melda.info/pktriggercord_android.html)

## Cameras

- Pentax K-x
- Pentax K10D (Samsung GX-10) _<sub>requires firmware 1.20 or later</sub>_
- Pentax K20D (Samsung GX-20)
- Pentax K200D
- Pentax K-7
- Pentax K-r
- Pentax K-5
- Pentax K-m / K2000
- Pentax K-30
- Pentax K-01 (bugs)
- Pentax K-5 II / K-5 IIs
- Pentax K-50 (bugs)
- Pentax K-3 / K-3 II
- Pentax K-500

### Limited support

- Pentax istDS2
- Pentax istDL
- Pentax K100D Super

### Not working

- Pentax istD
- Pentax istDS
- Pentax K110D
- Pentax K100D
- Pentax K-S2

### Testing

*The following cameras are not yet tested, please contact us if you can help.*

- Pentax K-S1

## More info

You can find more info about the project at http://pktriggercord.melda.info

More info about the Pentax SCSI protocol can be found here: https://github.com/asalamon74/pktriggercord/blob/master/pentax_scsi_protocol.md


Changelog
version 0.85.01
	RAD10 compilation (thx FrankBijnen)
	Exposure mode conversion bugfix (thx blemasle)
	Testing status reader for KP
	Settings reading for K5II (thx blemasle)
	remote_bulb_mode_press_press setting (thx blemasle)
	Updating lens info database ( exiftool 11.99 )
	Testing support for 645Z
	Better windows cross-compliation (thx blemasle)
	New servermode subcommands: set_buffer_type, get_buffer_type, set_aperture (thx Yale Kozinski)
	Add file_num_start option (thx colbyyh2)
	Testing K-3III
	Code cleanups (thx jirislaby)

version 0.85.00 ( 2019-JAN-27 )
	K-1 battery fields fix
	K-70: read one push bracketing field
	Using one push bracketing field in command line
	Print out command line in debug mode
	K-70 bulb mode bugfix
	--settings
	K-1 bulb mode bugfix (testing)
	--dangerous no longer required for K-70
	bulb_mode setting reading for K-70, K-50, K-01
	better handling of unknown and n/a settings
	K-x settings parser (very simple)
	Checking both both /dev/block and /dev
	Android: using gradle instead of ant
	Android: nicer image download indicator
	GUI ISO table bugfix
	Updating lens info database ( exiftool 10.70 )
	Android: better support of landscape mode
	GTK code cleanup
	Requires at least gtk+ 2.18
	GTK status info/status hex bugfix
	Using command 0x02 0x00 for buffer mask (K-70, K-3II)
	Possibility to specify stdout as output file using -
	Default mode prints out status info instead of taking a picture
	using_aperture_ring setting reading for K-70
	shake_reduction setting reading for K-70
	astrotracer setting reading for K70, K-3II
	horizon_correction setting reading for K-70
	astrotracer timer sec setting reading for K-3II
	K-3II bulb mode bugfix (testing)
	Restore bulb_timer, astrotracer fields after bulb
	CLI: Support for single buffer mode mode (K-70)
	GUI: Settings info
	GUI: Support for single buffer mode mode (K-70)
	Testing support for KP
	K-70 GUI image save bugfixes
	Android: Better error message if root access is missing.
	GUI ISO table simplification
	GUI Preview image bugfix
	Filename extension bugfixes
	Avoid double unrecognized option error message
	Show error message if servermode is specified in Windows
	Human readable selected/focused af point (for 11 AF points)
	K-70 AF selection mode reading hack
	K-70 reading selected af point
	--select_af_point also accepts raw numerical values
	Testing support for K-1 II

version 0.84.04 ( 2017-05-13 )
	Android: adding icon
	about/version code cleanup
	Testing suport for K-70
	K-70, K-1 selected_af_point hack
	Android: about window
	K-1 bufmask testing
	--dangerous hack
	GUI: Status hexdump
	New status field: auto_bracket_picture_counter (thx jmozmoz)
	GUI: Editable bulb exposure time
	GUI: Hide AF points if camera has more than 11
	K-S1 bugfix
	K-70 battery field reading bugfix
	Handling two new jpeg image tones: CrossProcessing, Flat, Auto
	Handling new white balance modes: MultiAuto, Manual2, Manual3, Kelvin1, Kelvin2, Kelvin3
	--read_datetime
	--read_firmware_version
	Fixing incorrectly recognized exposure submodes
	Handling new exposure submodes: Forest, NightSceneHDR, BacklightSilhouette, BlueSky
	Renaming exposure submode to scene mode
	--settings_hex
	Avoid infinite loop if green button fails
	Adding bufmask field to --status printing
	Handling higher ISO values in the GUI
	GUI: Avoid thumbnail reading, downscale preview instead
	--dump_memory

version 0.84.02 ( 2016-10-31 )
	Android code cleanup
	Possibility to read lens database from latest exiftool
	Updating lens info database ( exiftool 10.300 )
	Testing support for K-1
	Small speedup
	Increase Android 5 compatibility
	Android version requires Android 4.1 (http://stackoverflow.com/a/26422855/21348 )

version 0.84.00 ( 2016-05-09 )
	servermode subcommand: set_shutter_speed, set_iso, usleep
	servermode: avoid segmentation fault if camera is not connected
	android: bracketing (if set on the camera)
	openbsd port ( thx Hannu Vuolasaho )
	K-500 support
	Android preferences: showPreview, saveDir
	Android cheching directory
	Changing android default directory to /mnt/sdcard

version 0.83.00 ( 2016-01-29 )
	Limited K-3 II support (Testing)
	Makefile cleanup (exiftool)
	K-3 support ( thx Tao Wang )
	Eliminate GTK settings window closing warning
	Limited K-S1, K-S2 support ( thx Tao Wang )
	Eliminate segmentation fault (non-root Linux usage)
	autosave, save as dialog cancel bugfix
	Linux/Android: driveinfo bugfix
	Android file download progress/speedup

version 0.82.04 ( 2015-08-28 )
	More debug messages
	jpeg contrast/hue/sharpness/saturation GUI bugfix
	Image tone bugfix (thx Mike Protts)
	Disable muted jpeg image tone for K-x because it's not working correctly
	Android build fixes
	Disable jpeg hue change for old cameras without jpeg hue setting
	Makefile code cleanup
	Project moved to github.com
	Update/addition of dlls to windows port
	Servermode documentation

version 0.82.02 ( 2015-06-28 )
	--pentax_debug_mode ( thx Samo Penic )
	Testing support for K-50
	Updating lens info database ( exiftool 9.700 )
	Jpeg quality bugfix (cmd,gui)
	Better model recognition

version 0.82.00 ( 2014-11-01 )
	Using newer mingw for Windows crosscompile
	Attemp to fix root requirement
	strncpy bugfix
	Also check /dev/block if device file is not found in /dev
	android port of pktriggercord-cli
	pktriggercord-cli:
	  --servermode ( not for Windows )
	  --servermode_timeout
	servermode subcommands:
	    stopserver, disconnect, echo, connect, update_status,
	    get_camera_name, get_lens_name, get_current_shutter_speed,
	    get_current_aperture, get_current_iso, focus, get_bufmask,
	    get_preview_buffer, get_buffer, delete_buffer
	android GUI: status reading, focus, single shutter, multi shutter

version 0.81.04 ( 2014-05-01 )
	--noshutter ( thx Piotr Dlugosz )
	Support for Samsung GX-1L ( not working )
	K-5 IIs jpeg star calculation fix ( commandline )
	Unknown camera segmentation fault bugfix
	Limited support for K-3 ( not working )

version 0.81.02 ( 2013-11-01 )

pslr_enum.c
/*
    pkTriggerCord
    Copyright (C) 2011-2019 Andras Salamon <andras.salamon@melda.info>
    Remote control of Pentax DSLR cameras.

    Support for K200D added by Jens Dreyer <jens.dreyer@udo.edu> 04/2011
    Support for K-r added by Vincenc Podobnik <vincenc.podobnik@gmail.com> 06/2011
    Support for K-30 added by Camilo Polymeris <cpolymeris@gmail.com> 09/2012
    Support for K-01 added by Ethan Queen <ethanqueen@gmail.com> 01/2013
    Support for K-3 added by Tao Wang <twang2218@gmail.com> 01/2016

    based on:

    PK-Remote
    Remote control of Pentax DSLR cameras.
    Copyright (C) 2008 Pontus Lidman <pontus@lysator.liu.se>

    PK-Remote for Windows
    Copyright (C) 2010 Tomasz Kos

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.

    You should have received a copy of the GNU General Public License
    and GNU Lesser General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
#include "pslr_enum.h"

#include <string.h>
#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>

const char* pslr_color_space_str[PSLR_COLOR_SPACE_MAX] = {
    "sRGB",
    "AdobeRGB"
};

const char* pslr_af_mode_str[PSLR_AF_MODE_MAX] = {
    "MF",
    "AF.S",
    "AF.C",
    "AF.A"
};

const char* pslr_ae_metering_str[PSLR_AE_METERING_MAX] = {
    "Multi",
    "Center",
    "Spot"
};

const char* pslr_flash_mode_str[PSLR_FLASH_MODE_MAX] = {
    "Manual",
    "Manual-RedEye",
    "Slow",
    "Slow-RedEye",
    "TrailingCurtain",
    "Auto",
    "Auto-RedEye",
    "TrailingCurtain", /* maybe in manual mode?? */
    "Wireless"
};

const char* pslr_drive_mode_str[PSLR_DRIVE_MODE_MAX] = {
    "Single", /* Bracketing also returns Single */
    "Continuous-HI",
    "SelfTimer-12",
    "SelfTimer-2",
    "Remote",
    "Remote-3",
    "Continuous-LO"
};

const char*  pslr_af_point_sel_str[PSLR_AF_POINT_SEL_MAX] = {
    "Auto-5",
    "Select",
    "Spot",
    "Auto-11",
    "Expanded"
};

const char*  pslr_af11_point_str[11] = {
    "topleft",
    "topmiddle",
    "topright",
    "farleft",
    "middleleft",
    "middlemiddle",
    "middleright",
    "farright",
    "bottomleft",
    "bottommiddle",
    "bottomright"
};

const char* pslr_jpeg_image_tone_str[PSLR_JPEG_IMAGE_TONE_MAX] = {
    "Natural",
    "Bright",
    "Portrait",
    "Landscape",
    "Vibrant",
    "Monochrome",
    "Muted",
    "ReversalFilm",
    "BleachBypass",
    "Radiant",
    "CrossProcessing",
    "Flat",
    "Auto"
};

const char* pslr_white_balance_mode_str[PSLR_WHITE_BALANCE_MODE_MAX] = {
    "Auto",
    "Daylight",
    "Shade",
    "Cloudy",
    "Fluorescent_D",
    "Fluorescent_N",
    "Fluorescent_W",
    "Tungsten",
    "Flash",
    "Manual", /* sometimes called Manual1 */
    "Manual2",
    "Manual3",
    "Kelvin1",
    "Kelvin2",
    "Kelvin3",
    "Fluorescent_L",
    "CTE",
    "MultiAuto"
};

const char* pslr_custom_ev_steps_str[PSLR_CUSTOM_EV_STEPS_MAX] = {
    "1/2",
    "1/3"
};

const char* pslr_custom_sensitivity_steps_str[PSLR_CUSTOM_SENSITIVITY_STEPS_MAX] = {
    "1",
    "As"
};

const char* pslr_image_format_str[PSLR_IMAGE_FORMAT_MAX] = {
    "JPEG",
    "RAW",
    "RAW+"
};

const char* pslr_raw_format_str[PSLR_RAW_FORMAT_MAX] = {
    "PEF",
    "DNG"
};

const char* pslr_scene_mode_str[PSLR_SCENE_MODE_MAX] = {
    "NONE",
    "HISPEED",
    "DOF",
    "MTF",
    "STANDARD",
    "PORTRAIT",
    "LANDSCAPE",
    "MACRO",
    "SPORT",
    "NIGHTSCENEPORTRAIT",
    "NOFLASH",
    "NIGHTSCENE",
    "SURFANDSNOW",
    "TEXT",
    "SUNSET",
    "KIDS",
    "PET",
    "CANDLELIGHT",
    "MUSEUM",
    "19",
    "FOOD",
    "STAGE",
    "NIGHTSNAP",
    "SWALLOWDOF",
    "24",
    "NIGHTSCENEHDR",
    "BLUESKY",
    "FOREST",
    "28",
    "BLACKLIGHTSILHOUETTE"
};


/* case insenstive comparison - strnicmp */
int str_comparison_i (const char *s1, const char *s2, int n) {
    if ( s1 == NULL ) {
        return s2 == NULL ? 0 : -(*s2);
    }
    if (s2 == NULL) {
        return *s1;
    }

    char c1='\0', c2='\0';
    int length=0;
    while ( length<n && (c1 = tolower (*s1)) == (c2 = tolower (*s2))) {
        if (*s1 == '\0') {
            break;
        }
        ++s1;
        ++s2;
        ++length;
    }
    return c1 - c2;
}

int find_in_array( const char** array, int length, char* str ) {
    int i;
    int found_index=-1;
    size_t found_index_length=0;
    size_t string_length;
    for ( i = 0; i<length; ++i ) {
        string_length = strlen(array[i]);
        if ( (str_comparison_i( array[i], str, string_length ) == 0) && (string_length > found_index_length) ) {
            found_index_length = string_length;
            found_index = i;
        }
    }
    return found_index;
}

static
const char *get_pslr_str( const char** array, int length, int value ) {
    if (value >=0 && value < length) {
        return array[value];
    } else {
        char *ret = malloc(128);
        sprintf (ret, "Unknown value: %d", value);
        return ret;
    }
}


pslr_color_space_t pslr_get_color_space( char *str ) {
    return find_in_array( pslr_color_space_str, sizeof(pslr_color_space_str)/sizeof(pslr_color_space_str[0]),str);
}

const char *pslr_get_color_space_str( pslr_color_space_t value ) {
    return get_pslr_str( pslr_color_space_str, sizeof(pslr_color_space_str)/sizeof(pslr_color_space_str[0]),value);
}

pslr_af_mode_t pslr_get_af_mode( char *str ) {
    return find_in_array( pslr_af_mode_str, sizeof(pslr_af_mode_str)/sizeof(pslr_af_mode_str[0]),str);
}

const char *pslr_get_af_mode_str( pslr_af_mode_t value ) {
    return get_pslr_str( pslr_af_mode_str, sizeof(pslr_af_mode_str)/sizeof(pslr_af_mode_str[0]),value);
}

pslr_ae_metering_t pslr_get_ae_metering( char *str ) {
    return find_in_array( pslr_ae_metering_str, sizeof(pslr_ae_metering_str)/sizeof(pslr_ae_metering_str[0]),str);
}

const char *pslr_get_ae_metering_str( pslr_ae_metering_t value ) {
    return get_pslr_str( pslr_ae_metering_str, sizeof(pslr_ae_metering_str)/sizeof(pslr_ae_metering_str[0]),value);
}

pslr_flash_mode_t pslr_get_flash_mode( char *str ) {
    return find_in_array( pslr_flash_mode_str, sizeof(pslr_flash_mode_str)/sizeof(pslr_flash_mode_str[0]),str);
}

const char *pslr_get_flash_mode_str( pslr_flash_mode_t value ) {
    return get_pslr_str( pslr_flash_mode_str, sizeof(pslr_flash_mode_str)/sizeof(pslr_flash_mode_str[0]),value);
}

pslr_drive_mode_t pslr_get_drive_mode( char *str ) {
    return find_in_array( pslr_drive_mode_str, sizeof(pslr_drive_mode_str)/sizeof(pslr_drive_mode_str[0]),str);
}

const char *pslr_get_drive_mode_str( pslr_drive_mode_t value ) {
    return get_pslr_str( pslr_drive_mode_str, sizeof(pslr_drive_mode_str)/sizeof(pslr_drive_mode_str[0]),value);
}

pslr_af_point_sel_t pslr_get_af_point_sel( char *str ) {
    return find_in_array( pslr_af_point_sel_str, sizeof(pslr_af_point_sel_str)/sizeof(pslr_af_point_sel_str[0]),str);
}

const char *pslr_get_af_point_sel_str( pslr_af_point_sel_t value ) {
    return get_pslr_str( pslr_af_point_sel_str, sizeof(pslr_af_point_sel_str)/sizeof(pslr_af_point_sel_str[0]),value);
}

char *pslr_get_af11_point_str( uint32_t value ) {
    if (value==0) {
        return "none";
    }
    int bitidx=0;
    char *ret = malloc(1024);
    sprintf(ret, "%s", "");
    while (value>0 && bitidx<11) {
        if ((value & 0x01) == 1) {
            sprintf(ret, "%s%s%s", ret, strlen(ret) == 0 ? "" : ",", pslr_af11_point_str[bitidx]);
        }
        value >>= 1;
        ++bitidx;
    }
    if (value>0) {
        sprintf(ret, "%s", "invalid");
    }
    return ret;
}


pslr_jpeg_image_tone_t pslr_get_jpeg_image_tone( char *str ) {
    return find_in_array( pslr_jpeg_image_tone_str, sizeof(pslr_jpeg_image_tone_str)/sizeof(pslr_jpeg_image_tone_str[0]),str);
}

const char *pslr_get_jpeg_image_tone_str( pslr_jpeg_image_tone_t value ) {
    return get_pslr_str( pslr_jpeg_image_tone_str, sizeof(pslr_jpeg_image_tone_str)/sizeof(pslr_jpeg_image_tone_str[0]),value);
}

pslr_white_balance_mode_t pslr_get_white_balance_mode( char *str ) {
    return find_in_array( pslr_white_balance_mode_str, sizeof(pslr_white_balance_mode_str)/sizeof(pslr_white_balance_mode_str[0]),str);
}

const char *pslr_get_white_balance_mode_str( pslr_white_balance_mode_t value ) {
    return get_pslr_str( pslr_white_balance_mode_str, sizeof(pslr_white_balance_mode_str)/sizeof(pslr_white_balance_mode_str[0]),value);
}

const char *pslr_get_custom_ev_steps_str( pslr_custom_ev_steps_t value ) {
    return get_pslr_str( pslr_custom_ev_steps_str, sizeof(pslr_custom_ev_steps_str)/sizeof(pslr_custom_ev_steps_str[0]),value);
}

const char *pslr_get_custom_sensitivity_steps_str( pslr_custom_sensitivity_steps_t value ) {
    return get_pslr_str( pslr_custom_sensitivity_steps_str, sizeof(pslr_custom_sensitivity_steps_str)/sizeof(pslr_custom_sensitivity_steps_str[0]),value);
}

const char *pslr_get_image_format_str( pslr_image_format_t value ) {
    return get_pslr_str( pslr_image_format_str, sizeof(pslr_image_format_str)/sizeof(pslr_image_format_str[0]),value);
}

const char *pslr_get_raw_format_str( pslr_raw_format_t value ) {
    return get_pslr_str( pslr_raw_format_str, sizeof(pslr_raw_format_str)/sizeof(pslr_raw_format_str[0]),value);
}

const char *pslr_get_scene_mode_str( pslr_scene_mode_t value ) {
    return get_pslr_str( pslr_scene_mode_str, sizeof(pslr_scene_mode_str)/sizeof(pslr_scene_mode_str[0]),value);
}
